pyro/ops/newton.py

```diff
@@ -1,23 +1,83 @@
 from __future__ import absolute_import, division, print_function
 
+import numpy as np
+
 import torch
 from torch.autograd import grad
 
 from pyro.util import warn_if_nan
 
 
-def newton_step_2d(loss, x, trust_radius=None):
+def _determinant_3d(H):
     """
-    Performs a Newton update step to minimize loss on a batch of 2-dimensional
-    variables, optionally regularizing to constrain to a trust region.
+    Returns the determinants of a batched 3-D matrix
+    """
+    diag_terms = H[..., 0, 0] * H[..., 1, 1] * H[..., 2, 2]
+    diag_terms = diag_terms + H[..., 0, 1] * H[..., 1, 2] * H[..., 2, 0]
+    diag_terms = diag_terms + H[..., 1, 0] * H[..., 2, 1] * H[..., 0, 2]
+    antidiag_terms = H[..., 2, 0] * H[..., 1, 1] * H[..., 0, 2]
+    antidiag_terms = antidiag_terms + H[..., 0, 1] * H[..., 1, 0] * H[..., 2, 2]
+    antidiag_terms = antidiag_terms + H[..., 0, 0] * H[..., 2, 1] * H[..., 1, 2]
+    detH = diag_terms - antidiag_terms
+    return detH
+
+
+def _eig_3d(H):
+    """
+    Returns the eigenvalues of a symmetric batched 3-D matrix
+    """
+    p1 = H[..., 0, 1].pow(2) + H[..., 0, 2].pow(2) + H[..., 1, 2].pow(2)
+    q = (H[..., 0, 0] + H[..., 1, 1] + H[..., 2, 2]) / 3
+    p2 = (H[..., 0, 0] - q).pow(2) + (H[..., 1, 1] - q).pow(2) + (H[..., 2, 2] - q).pow(2) + 2 * p1
+    p = torch.sqrt(p2 / 6)
+    if q.shape == ():
+        B = (1 / p) * (H - q * torch.eye(3))
+    else:
+        B = (1 / p).unsqueeze(-1).unsqueeze(-1) * (H - q.unsqueeze(-1).unsqueeze(-1) * torch.eye(3))
+    r = _determinant_3d(B) / 2
+    phi = r.acos().unsqueeze(-1).unsqueeze(-1).expand(r.shape + (3, 3)) / 3
```

Comment:
nit: make this cheaper by moving the division op inside the expand
```py
phi = (r.acos() / 3).unsqueeze(-1).unsqueeze(-1).expand(r.shape + (3, 3))
```

Reply:
done